src/Renci.SshNet/RemotePathQuoteTransformation.cs

```diff
@@ -0,0 +1,161 @@
+﻿using System;
+using System.Text;
+
+namespace Renci.SshNet
+{
+    /// <summary>
+    /// Quotes a path in a way to be suitable to be used with a shell.
+    /// </summary>
+    internal class RemotePathQuoteTransformation : IRemotePathTransformation
+    {
+        /// <summary>
+        /// Quotes a path in a way to be suitable to be used with a shell.
+        /// </summary>
+        /// <param name="path">The path to transform.</param>
+        /// <returns>
+        /// A quoted path.
+        /// </returns>
+        /// <exception cref="ArgumentNullException"><paramref name="path"/> is <c>null</c>.</exception>
+        /// <remarks>
+        /// <para>
+        /// If <paramref name="path"/> contains a single-quote, that character is embedded
+        /// in quotation marks (eg. "'"). Sequences of single-quotes are grouped in a single
+        /// pair of quotation marks.
+        /// </para>
+        /// <para>
+        /// If a shell command contains an exclamation mark (!), the C-Shell interprets it as a
+        /// meta-character for history substitution. This even works inside single-quotes or
+        /// quotation marks, unless escaped with a backslash (\).
+        /// </para>
+        /// <para>
+        /// References:
+        /// <list type="bullet">
+        ///   <item>
+        ///     <description><a href="http://pubs.opengroup.org/onlinepubs/7908799/xcu/chap2.html">Shell Command Language</a></description>
+        ///   </item>
+        ///   <item>
+        ///     <description><a href="https://earthsci.stanford.edu/computing/unix/shell/specialchars.php">Unix C-Shell special characters and their uses</a></description>
+        ///   </item>
+        ///   <item>
+        ///     <description><a href="https://docstore.mik.ua/orelly/unix3/upt/ch27_13.htm">Differences Between Bourne and C Shell Quoting</a></description>
+        ///   </item>
+        /// </list>
+        /// </para>
+        /// </remarks>
+        /// <returns>
+        /// The transformed path.
+        /// </returns>
+        /// <exception cref="ArgumentNullException"><paramref name="path"/> is <c>null</c>.</exception>
+        public string Transform(string path)
+        {
+            if (path == null)
```

Comment:
Consider checking if the path doesn't consist of characters that don't need escaping (`[a-zA-Z0-9,._+:@%/-]`).
MicroTik and Cisco and others that don't do unescaping, will probably have paths that don't require escaping. So they will work even when using the Transform.

Next, you could scan the string for the single quote character (`'`). If it isn't present in the path, you can pre and postpend it to escape the path.

If the two cases above don't apply, scan through the string and prepend every char with a backslash except those that don't require escaping. There is no need to enclose the entire string.

examples:
```
file1.txt     -> file1.txt
my file.txt   -> 'my file.txt'
my 'file'.txt -> my\ \'file\'.txt
```
Also for `!` this is fine, e.g.:
```
/tmp$ echo text
text
/tmp$ echo !-1
echo echo text
echo text
/tmp$ echo \!-1
!-1
```

Reply:
I'll give it some more thought.

If we only enclose characters in single quotes when we encounter "special characters", then - on SCP servers that do not unescape - we risk having **SSH.NET** work just fine until a file that does require escaping is uploaded or downloaded.

I would prefer our implementation to either always work out-of-the-box, or to fail immediately (and require disabling of quoting through the **RemotePathTransformation** property).

Reply:
> If we only enclose characters in single quotes when we encounter "special characters", then - on SCP servers that do not unescape - we risk having SSH.NET work just fine until a file that does require escaping is uploaded or downloaded.

It's a bit hard to know how to deal with a file that requires escaping on an scp server that doesn't unescape. So servers that don't unescape (on embedded devices) are not likely to have files that need escaping.

With the current implementation when downloading a file you observe the failure because it can't find the file. If you were to upload a file, you may end up creating a file called `'firmware'` instead of `firmware`.

Reply:
I agree that - with the current implementation -, you'll only get failures when downloading a file, and you'll indeed risk uploading files enclosed in single quotes (and/or with double quotes, or a backslash). 

However, the advantage of the current implementation is that this behavior will immediately manifest itself and not only upon the first file that contains characters that deem to require quoting. Nothing bothers me more than something that works "most of the time".

PS. I only tested a few, but AFAIK none of the Windows SCP servers unescape even though you'll definitely have files or directories that contain "special characters" (mostly spaces).